File: src/core/tests/invariants/combat.ts
// 전투 상태 불변식 검사기 (INVARIANTS-FROM-CLAUDE-CODE.md Part 1).
// 매 step 후(또는 임의 시점) GameState를 받아 위반 목록을 반환. 순수·무throw.
// 관측 경계에서만 판정(8.2) — units·roundOrder·cursor·phase + buildObservation 투영.
import type { GameState } from "../../types.ts";
import { buildObservation } from "../../engine.ts";
import { STATUS_DEFS } from "../../../data/statuses.ts";
import { Violations, type Violation } from "./types.ts";

const ACTION_CONST = 10000; // turnOrder.ts와 동일(서열 = ACTION_CONST/speed)

export function checkCombatInvariants(s: GameState): Violation[] {
  const v = new Violations();

  // ── A. 유닛 상태 무결성 ──
  const seen = new Set<string>();
  for (const u of s.units) {
    v.check(u.hp >= 0, "A1", "CRIT", () => `${u.uid} hp<0 (${u.hp})`);
    v.check(u.hp <= u.hpMax, "A2", "CRIT", () => `${u.uid} hp>hpMax (${u.hp}/${u.hpMax})`);
    v.check(u.shield >= 0, "A3", "NORM", () => `${u.uid} shield<0 (${u.shield})`);
    v.check(u.alive === u.hp > 0, "A4", "CRIT", () => `${u.uid} alive↔hp 불일치 (alive=${u.alive}, hp=${u.hp})`);
    v.check(!seen.has(u.uid), "A6", "CRIT", () => `uid 중복: ${u.uid}`);
    seen.add(u.uid);
    v.check(u.activeSkillIds.length <= 4, "A8", "NORM", () => `${u.uid} activeSkillIds>4 (${u.activeSkillIds.length})`);
    // C9 쿨다운 비음수
    for (const id of Object.keys(u.cooldowns)) {
      v.check(u.cooldowns[id] >= 0, "C9", "NORM", () => `${u.uid} 쿨다운<0 (${id}=${u.cooldowns[id]})`);
    }
    // E (참조 무결성): 모든 상태 defId가 정의에 존재
    for (const st of u.statuses) {
      v.check(!!STATUS_DEFS[st.defId], "E-ref", "CRIT", () => `${u.uid} 미정의 상태 defId=${st.defId}`);
    }
  }

  // ── C. 턴/라운드 서열 ──
  const normals = s.roundOrder.filter((e) => e.kind === "normal");
  for (const e of normals) {
    v.check(e.speed >= 1, "C3", "CRIT", () => `정규 엔트리 speed<1 (${e.uid}=${e.speed})`);
  }
  // C2: 정규 엔트리 서열 정렬(ACTION_CONST/speed 오름차순, 동점 uid 사전순) — roundOrder 내 정규 부분열이 정렬돼야
  for (let i = 1; i < normals.length; i++) {
    const a = normals[i - 1];
    const b = normals[i];
    const av = ACTION_CONST / a.speed;
    const bv = ACTION_CONST / b.speed;
    const ordered = av < bv || (av === bv && a.uid < b.uid) || (av === bv && a.uid === b.uid);
    v.check(ordered, "C2", "CRIT", () => `정규 서열 정렬 위반: ${a.uid}(${av}) 다음 ${b.uid}(${bv})`);
  }
  // C5/C7: current가 있으면 roundOrder[cursor]와 동일 ∧ 그 유닛 생존
  if (s.current !== null) {
    const atCursor = s.roundOrder[s.cursor];
    v.check(atCursor === s.current, "C7", "CRIT", () => `current ≠ roundOrder[cursor] (cursor=${s.cursor})`);
    const u = s.units.find((x) => x.uid === s.current!.uid);
    v.check(!!u && u.alive, "C5", "CRIT", () => `current가 죽은/없는 유닛 (${s.current!.uid})`);
  }

  // ── J. 관측 충실성(8.2) — buildObservation이 상태를 그대로 투영 ──
  const obs = buildObservation(s);
  v.check(obs.round === s.round, "J1", "CRIT", () => `obs.round≠state.round (${obs.round}/${s.round})`);
  v.check(obs.phase === s.phase, "J1", "CRIT", () => `obs.phase≠state.phase (${obs.phase}/${s.phase})`);
  v.check(obs.cursorIndex === s.cursor, "J1", "CRIT", () => `obs.cursorIndex≠state.cursor (${obs.cursorIndex}/${s.cursor})`);
  v.check(obs.order.length === s.roundOrder.length, "J1", "CRIT", () => `obs.order 길이≠roundOrder (${obs.order.length}/${s.roundOrder.length})`);
  v.check(
    obs.allies.length + obs.enemies.length === s.units.length,
    "J2", "NORM", () => `allies+enemies≠units (${obs.allies.length}+${obs.enemies.length}/${s.units.length})`,
  );

  return v.list;
}

File: engine/spr-types/src/combat.rs
//! 전투 런타임 타입 (TS `types/runtime.ts`). 이벤트는 canonical 직렬화 대상(SERIALIZATION-CONTRACT).
use crate::data::Pos;
use crate::rng::Rng;
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

/// 컴파일된 패시브 룰 — 발동 카운터 포함(턴/전투당 캡). TS CompiledRule.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CompiledRule {
    pub rule: crate::passives::PassiveRule,
    pub via_kind: String, // "skill" | "trait" | "node"
    pub via_id: String,
    pub idx: i64,
    pub fired_this_turn: i64,
    pub fired_this_battle: i64,
}

/// 상태이상 인스턴스 (원장 1건).
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct StatusInstance {
    pub def_id: String,
    pub stacks: i64,
    pub duration: i64,
    pub source_uid: String,
    pub source_skill_id: Option<String>,
}

/// 전투 중 유닛. (rules/statMods 등은 패시브 슬라이스서 확장)
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Unit {
    pub uid: String,
    pub side: String, // "ally" | "enemy"
    pub char_id: String,
    pub name: String,
    pub pos: Pos,
    pub hp_max: i64,
    pub hp: i64,
    pub shield: i64,
    pub speed_min: i64,
    pub speed_max: i64,
    pub evasion: i64,
    pub accuracy: i64,
    pub crit_chance: i64,
    pub crit_multiplier: i64,
    pub active_skill_ids: Vec<String>,
    pub cooldowns: HashMap<String, i64>,
    pub statuses: Vec<StatusInstance>,
    pub alive: bool,
    pub stat_mods: HashMap<String, i64>,
    pub turn_count: i64,
    pub skill_dmg_bonus: HashMap<String, i64>, // 스킬별 데미지 보너스(런 성장, 4.6)
    pub rules: Vec<CompiledRule>,   // 컴파일된 패시브(스킬 passives + 특성 + 노드)
    pub ai_profile_id: Option<String>, // 적/자동플레이 행동결정 정책(data/ai)
    pub equip_dmg_flat: i64,        // 장착 무기 dmgFlat 합(4.3) — computeDamage 합산
    pub equip_shield_gain_add: i64, // 장착 방어구 쉴드획득 보정 합(4.3)
    /// 소환수(R2) 여부 — 만료 시 소멸. 기본 false면 직렬화 생략(골든 무변).
    #[serde(default, skip_serializing_if = "is_false")]
    pub summoned: bool,
    /// 소환 만료 라운드(R2) — round > expires_round면 라운드 시작 시 소멸. 기본 0면 직렬화 생략.
    #[serde(rename = "expiresRound", default, skip_serializing_if = "is_zero_i64")]
    pub expires_round: i64,
}

fn is_false(b: &bool) -> bool { !*b }
fn is_zero_i64(n: &i64) -> bool { *n == 0 }

pub type TurnKind = &'static str; // "normal" | "interrupt"

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct QueueEntry {
    pub uid: String,
    pub kind: String,
    pub speed: i64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SpeedRoll {
    pub uid: String,
    #[serde(rename = "speedMin")]
    pub speed_min: i64,
    #[serde(rename = "speedMax")]
    pub speed_max: i64,
    pub roll: i64,
    #[serde(rename = "speedMod")]
    pub speed_mod: i64,
    pub speed: i64,
}

/// 플레이어/AI 행동(런타임 입력 — step이 소비). 행동벡터로 기록·재생(TS Action).
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type")]
pub enum Action {
    #[serde(rename = "skill")]
    Skill {
        #[serde(rename = "skillId")]
        skill_id: String,
        #[serde(rename = "targetUid", skip_serializing_if = "Option::is_none", default)]
        target_uid: Option<String>,
        #[serde(rename = "targetCell", skip_serializing_if = "Option::is_none", default)]
        target_cell: Option<Pos>,
        #[serde(skip_serializing_if = "Option::is_none", default)]
        cells: Option<Vec<Pos>>,
    },
    #[serde(rename = "skip")]
    Skip,
}

/// 이벤트 로그 (canonical 직렬화 — TS GameEvent). 변종은 포팅 진행하며 추가.
#[derive(Debug, Clone, Serialize)]
#[serde(tag = "t")]
pub enum GameEvent {
    #[serde(rename = "roundStart")]
    RoundStart {
        round: i64,
        order: Vec<QueueEntry>,
        rolls: Vec<SpeedRoll>,
    },
    #[serde(rename = "turnStart")]
    TurnStart { uid: String, kind: String },
    #[serde(rename = "damage")]
    Damage {
        #[serde(rename = "targetUid")]
        target_uid: String,
        base: i64,
        #[serde(rename = "final")]
        final_: i64,
        #[serde(rename = "toShield")]
        to_shield: i64,
        #[serde(rename = "toHp")]
        to_hp: i64,
    },
    #[serde(rename = "death")]
    Death { uid: String },
    #[serde(rename = "statusApplied")]
    StatusApplied {
        #[serde(rename = "targetUid")]
        target_uid: String,
        #[serde(rename = "statusId")]
        status_id: String,
        stacks: i64,
        duration: i64,
    },
    #[serde(rename = "heal")]
    Heal {
        #[serde(rename = "targetUid")]
        target_uid: String,
        amount: i64,
    },
    #[serde(rename = "statusTick")]
    StatusTick {
        #[serde(rename = "targetUid")]
        target_uid: String,
        #[serde(rename = "statusId")]
        status_id: String,
        dmg: i64,
    },
    #[serde(rename = "skillUsed")]
    SkillUsed {
        uid: String,
        #[serde(rename = "skillId")]
        skill_id: String,
        #[serde(rename = "targetUid", skip_serializing_if = "Option::is_none")]
        target_uid: Option<String>,
    },
    #[serde(rename = "move")]
    Move { uid: String, from: Pos, to: Pos },
    #[serde(rename = "cleanse")]
    Cleanse {
        #[serde(rename = "targetUid")]
        target_uid: String,
    },
    #[serde(rename = "shieldGain")]
    ShieldGain {
        #[serde(rename = "targetUid")]
        target_uid: String,
        amount: i64,
    },
    #[serde(rename = "miss")]
    Miss {
        uid: String,
        #[serde(rename = "targetUid")]
        target_uid: String,
        chance: i64,
    },
    #[serde(rename = "hit")]
    Hit {
        uid: String,
        #[serde(rename = "targetUid")]
        target_uid: String,
        chance: i64,
        crit: bool,
    },
    #[serde(rename = "interrupt")]
    Interrupt { uid: String },
    #[serde(rename = "dialog")]
    Dialog {
        #[serde(skip_serializing_if = "Option::is_none")]
        speaker: Option<String>,
        text: String,
    },
    #[serde(rename = "battleEnd")]
    BattleEnd { phase: String },
    #[serde(rename = "skip")]
    Skip { uid: String, reason: String },
    /// 소환(R2) — 임시 아군 유닛 생성.
    #[serde(rename = "summon")]
    Summon {
        uid: String,
        #[serde(rename = "charId")]
        char_id: String,
    },
}

/// 전투 상태. 세이브 직렬화 대상 — `log`은 append-only(프로덕션 로직 미참조)라 skip(GameEvent serde 트리 회피, 복원 시 빈 벡터).
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct GameState {
    pub rng: Rng,
    pub round: i64,
    pub units: Vec<Unit>,
    pub round_order: Vec<QueueEntry>,
    pub cursor: i64,
    pub current: Option<QueueEntry>,
    pub phase: String, // "inProgress" | "allyWin" | "enemyWin"
    #[serde(skip)]
    pub log: Vec<GameEvent>,
    pub ally_formation: Option<crate::data::FormationLayout>,
    pub enemy_formation: Option<crate::data::FormationLayout>,
    // 패시브 재진입 가드(P0-3) — 후속 슬라이스서 사용
    pub fire_depth: i64,
    pub fire_active_keys: Vec<String>,
    /// 소환 템플릿(R2) — charId→사전빌드 Unit. 전투 생성 시 아군 스킬의 summon charId 스캔. 비면 직렬화 생략(골든 무변).
    #[serde(rename = "summonTemplates", default, skip_serializing_if = "std::collections::HashMap::is_empty")]
    pub summon_templates: std::collections::HashMap<String, Unit>,
}
